eclipse/src/saros/filesystem/checksum/FileContentNotifierBridge.java

```diff
@@ -51,8 +74,28 @@ public void resourceChanged(IResourceChangeEvent event) {
 
         final IFile file = delta.getResource().getAdapter(IFile.class);
 
-        for (IFileContentChangedListener listener : fileContentChangedListeners)
-          listener.fileContentChanged(ResourceAdapterFactory.create(file));
+        final ISarosSession sarosSession = sarosSessionManager.getSession();
```

Comment:
So you are basically breaking the cache implementation ... congratulation !

Does the interface states it operates only on SHARED files ? I do not think so.

Reply:
Start a session. Stop it. Modify some files that were part of the former sharing. Start a session again. You now have dirty checksums.. And if the dirty checksums match the remote checksums (which is likely) you will produce inconsistencies.

Reply:
I guess moving it into session scope can be pretty annoying if a session crashes for what reason. You should not assume that everyone is using SSDs or 16 Gigabytes of system memory (OS cache).

Reply:
Both solutions are possible and offer their own advantages.

Separating the negotiation factories removes the dirty circular dependency workaround (and removes another one currently necessary for the [`NegotiationFactory`](https://github.com/saros-project/saros/blob/master/core/src/saros/negotiation/NegotiationFactory.java#L58)). But, as you mentioned, it only provides the caching benefits during the running session (i.e. for additional resource negotiations with new participants).

Making the caching logic generic and still using Eclipse `IFile` for the Saros/E cache removes the reference point lookup from the notification bridge logic and still offers the caching benefit for starting new sessions.
But, the cache being session-independent puts us in the weird position that Saros could potentially still use up memory and computing power even though there is currently no Saros session running. Even if the user is not planning on starting another session, the current caching model from the last session is still kept in memory and updated constantly for every change in the workspace.

I implemented a rough draft for both approaches:
[Move checksum cache to session context](https://github.com/saros-project/saros/commit/3fbc9e55447baad4b71a546cdf417525161c1baf)
[Make filesystem checksum cache generic](https://github.com/saros-project/saros/commit/0162fb579738a07b853391b5cfadbe657c265dc3)

I personally would still tend towards making the caching logic session-bound. (Also, I don't plan on fixing this as part of this PR but rather as a separate PR.)

@srossbach @m273d15 @stefaus What are your opinions?

Reply:
I think moving the cache in the session context reduces the benefit and leads to the question "Do we need this cache?".

@srossbach Do you know the history of this cache (who requested the optimization and why)?

If we want to/should keep the cache I think we should use the second alternative (make the cache generic; even if I would like to remove the circular dependency). Otherwise, the benefit seems to be too small.

Reply:
I would also question the usefulness of the cache in general. If the user were to be working on huge projects, their hardware would have to match anyways as they otherwise could not load the project into the IDE or work on it in any sensible/timely manner.
But I am still a bit torn on this as I do see the utility during a session. Saros already hogs a lot of resources during a session, so actually using up more resources by creating checksums for all shared resources might push it over the edge (i.e. make the system noticeably slow down). On the other hand, the added cache still produces additional overhead during the session by keeping the checksums up-to-date, even if no further resource negotiation is planned, so removing it might benefit the general Saros performance (but not by a whole lot).

This would be a whole different discussion if the cache could also be used by the watchdog during the session as it would then serve a far broader purpose in this case. But this is currently not possible/not an option as the negotiation cache works on the file content on disc and the watchdog works on the editor content. Working on editor content is important for the watchdog to always work on the most current information, so using the content on disc is not an option. And using the editor content for the negotiation cache is not an option either (for now) as the negotiation could contain binary files which can't be opened in an editor. These caches could, however, be unified if we were to introduce the unified content handling (i.e. adding a binary check and always using the editor content for non-binary files, even for the negotiations).

If we want to keep the cache, I still disagree that moving it to the session context removes most of its benefits. I think it is much more likely that more participants are invited during the session than that the user is starting a new session on the same resources. (Yes, the session crashing/timing out is an exception to this, but I don't think this can be categorized as a default/common case.)

Also, as already mentioned, I am unhappy with the cache still running and using up system resources even though there is no Saros session running. In general, this is a very unexpected behavior from the users perspective and, if this were to negatively impact the system performance, could lead to lots of complaints/uninstalls.
Even if this is an extreme case that might not be that common (or even very uncommon), I would still question the benefits of the cache compared to its drawbacks, especially outside of the session. The overhead gets worse with increases in project size and decreases in hardware resources/power, so the users that would most benefit from this cache would also be impacted most negatively by its overhead.

Reply:
As mentioned in the standups I think it makes sense to move the cache into the session context. It is a tradeoff and not suitable for all cases, but as @tobous mentioned I don't think its good that Saros still updates checksums outside of sessions, burning resources without use. But keeping it inside the session context is good for multiple invites especially for Saros Server.

Reply:
The cache do not update checksums. It just invalidates existing stored checksums for files that were changed.

Reply:
> The cache do not update checksums. It just invalidates existing stored checksums for files that were changed.

You are right. I misread the code. But it still creates overhead with calculating the path hashes, but it isn't as bad as I thought.

I still think it is at its most useful during the session as I would assume it is more likely to start a session with multiple participants or add participants to a running session (where the cache would still work for all non-modified shared files) than starting another session on the same resources in the same application lifecycle.

Furthermore, starting a new session requires a break in the workflow anyways, so additional time to re-populate the cache should not be that much of an issue in my opinion.

Reply:
Resolved in #1072.